### PR TITLE
First homework

### PR DESCRIPTION
**Pull request by ivanseki**

### CODE REVIEW

File: android-vjestina-homework-movie-app-data-mock/app/build.gradle [hunk 1]
```diff
@@ -52,17 +52,19 @@ dependencies {
     implementation "androidx.core:core-ktx:$core_ktx_version"
     implementation 'androidx.appcompat:appcompat:1.5.1'
     implementation 'com.google.android.material:material:1.6.1'
-    implementation "androidx.compose.ui:ui:1.3.0-beta03"
-    implementation "androidx.compose.material:material:1.3.0-beta03"
+    //implementation "androidx.compose.ui:ui:1.3.0-beta03"
+    //implementation "androidx.compose.material:material:1.3.0-beta03"
+    implementation "androidx.compose.material3:material3-window-size-class:1.0.0-rc01"
+    implementation "androidx.compose.material3:material3:1.0.0-rc01"
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/build.gradle]: Please revert these changes. 
Everything can be implemented with the libraries provided.

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/ActorCard.kt [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    val character: String,
+)
+
+@Composable
+fun ActorCard(
+    item: ActorCardViewState,
+    modifier: Modifier = Modifier,
+) {
+    Card(
+        modifier = Modifier
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/ActorCard.kt]: The Card composable should use the `modifier` property which is passed into the composable function.

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/ActorCard.kt [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            .size(width = 125.dp, height = 209.dp)
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/ActorCard.kt]: Size should not be hardcoded here. The size of the component will be defined by the caller of this composable function by using the `modifier` property.

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/ActorCard.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            .clip(RoundedCornerShape(10.dp)),
+        elevation = CardDefaults.cardElevation(2.dp)
+    ) {
+        Column {
+            AsyncImage(
+                model = item.imageUrl,
+                contentDescription = item.name,
+                modifier = modifier
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/ActorCard.kt]: Should not use the modifier parameter. 
Create a new modifier by using `Modifier.xxx()`.
The modifier function param is only meant to be used in the root component so that the outside caller can change the look of the component itself.

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/ActorCard.kt [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+            Text(
+                text = item.name,
+                fontSize = 14.sp,
+                fontWeight = FontWeight.ExtraBold,
+                color = Color.Black,
+                modifier = modifier
+                    .padding(start = 5.dp, top = 5.dp)
+                    .width(87.dp)
+                    .height(35.dp)
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/ActorCard.kt]: Should not be hardcoded since the user can change the text size in the device settings thus rendering parts of the text invisible.

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/ActorCard.kt [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            )
+
+            Text(
+                text = item.character,
+                fontSize = 12.sp,
+                color = Color.Gray,
+                modifier = modifier
+                    .padding(start = 5.dp, top = 0.dp)
+                    .width(110.dp)
+                    .height(40.dp)
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/ActorCard.kt]: Should not be hardcoded since the user can change the text size in the device settings thus rendering parts of the text invisible.

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/UserScoreProgressBar.kt [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+) {
+    Box{
+        CircularProgressIndicator(
+            progress = score,
+            color = Color.Green,
+            strokeWidth = 4.dp,
+            modifier = modifier
+                .width(42.dp)
+                .height(42.dp),
+
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/UserScoreProgressBar.kt]: Redundant empty line

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/UserScoreProgressBar.kt [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.sp
+
+@Composable
+fun UserScoreProgressBar(
+    score: Float,
+    modifier: Modifier = Modifier
+) {
+    Box{
+        CircularProgressIndicator(
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/UserScoreProgressBar.kt]: Please implement the progress bar functionality using a `Canvas`. 
Becoming familiar with lower level drawing on the `Canvas` was one of the goals of this homework.

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/CrewItem.kt [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+import androidx.compose.ui.unit.sp
+
+data class CrewItemViewState(
+    val name: String,
+    val job: String,
+)
+
+@Composable
+fun CrewItem(
+    item: CrewItemViewState
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/CrewItem.kt]: Should also expose `modifier` as an input parameter to enable the caller of this function to customise its appearance.

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/FavoriteButton.kt [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+
+@Composable
+fun FavoriteButton(
+    modifier: Modifier = Modifier,
+    contentDescription: String = "Favourite button"
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/FavoriteButton.kt]: No need to provide this as an outside parameter. You can use `null` for `contentDescription`.

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/FavoriteButton.kt [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+) {
+    val favourite = remember { mutableStateOf(false) }
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/FavoriteButton.kt]: Favorite state should come from "the outside". Meaning as a function parameter. This is in general a good practice in compose to keep the components as dumb as possible and to cache the state somewhere else. 
You can read more about it here: https://developer.android.com/jetpack/compose/mental-model
Protip: `isFavorite` is a better name than `favorite` since it communicates what the data actually represents more clearly :)

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/FavoriteButton.kt [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+    Image(
+        painter = painterResource(id = if(favourite.value) R.drawable.ic_favorite_filled else R.drawable.ic_favorite_outlined),
+        contentDescription = contentDescription,
+        modifier = modifier
+            .clickable {
+                favourite.value = favourite.value.not()
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/FavoriteButton.kt]: Should invoke a callback function which should be passed to the function as a parameter

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/FavoriteButton.kt [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+            }
+            .size(width = 32.dp, height = 32.dp)
+            .background(Blue, CircleShape)
+            .padding(top = 8.dp, start = 5.dp, end = 5.dp, bottom = 5.dp)
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/FavoriteButton.kt]: Extract magic numbers to top level constants

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.layout.ContentScale
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import coil.compose.AsyncImage
+
+@Composable
+fun Moviecard(
+    item: Movie,
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt]: Should receive `MovieCardViewState` instead of `Movie`. 
Movie is a domain level object. With this design you are coupling the business with the UI layer. 
Try to imagine how this component would work in a different codebase with a different domain layer.

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    modifier: Modifier = Modifier,
+    onMovieItemClick: () -> Unit = {}
+) {
+    Card(
+        modifier = modifier
+            .width(122.dp)
+            .height(179.dp)
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt]: Width and Height should not be hardcoded. 
In general hardcoding both width and height is a bad idea since this makes the component fixed in size. In android the user can change the system text size in OS settings which would impact how much content fits into this component rendering some parts invisible or some text unreadable. 
Additionally, the component size should be handled by the caller of this function since the component is not aware of the context (in this case layout/screen) in which it will be used in.

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+            .clickable { onMovieItemClick() },
+        shape = RoundedCornerShape(10.dp),
+        elevation = CardDefaults.cardElevation(2.dp)
```
Comment (ivan-varga) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt:29]: Magic numbers

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+import androidx.compose.material3.CardDefaults
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.layout.ContentScale
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import coil.compose.AsyncImage
+
+@Composable
+fun Moviecard(
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt]: This composable should also expose `onLikeButtonClick` parameter which is a function that you can invoke when the user clicks on the favourite button.

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+        elevation = CardDefaults.cardElevation(2.dp)
+    ) {
+        Box {
+            AsyncImage(
+                model = item.imageUrl,
+                contentDescription = item.title,
+                contentScale = ContentScale.Crop
+            )
+
+            FavoriteButton(
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt:42]: Should also pass in the click listener to know when the button is clicked

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCategoryLabel.kt [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+)
+
+@Composable
+fun MovieCategoryLabel(
+    item: MovieCategoryLabelViewState,
+    modifier: Modifier = Modifier,
+) {
+    Column(modifier = modifier
+        .wrapContentWidth()
+        .height(22.dp)
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCategoryLabel.kt]: Hardcoding the height of the text is generally not a good idea since the text can be enlarged from OS settings. This solution would make the text partially visible and potentially unreadable. Use wrap content instead.

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCategoryLabel.kt [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+
+sealed class MovieCategoryLabelTextViewState()
+
+data class MovieCategoryAsString(val value: String) : MovieCategoryLabelTextViewState()
+
+data class MovieCategoryAsStringResource(@StringRes val value: Int) : MovieCategoryLabelTextViewState()
+
+data class MovieCategoryLabelViewState(
+    val itemId: Int,
+    var isSelected: Boolean,
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCategoryLabel.kt]: All ViewStates and their properties MUST be immutable.
When the user clicks a button, the action will be propagated through necessary layers (UI -> Presentation -> Business -> Data) and a new model will be given to the composable function with the state changed (this data will travel from the Data layer in the opposite direction). 
Change `var` to `val`.

**Review by ghost — CHANGES_REQUESTED** [review]
Please apply Android Studio formatting to all of your files. The shortcut on MacOS is cmd+alt+L and on win/linux is ctrl+alt+L

File: android-vjestina-homework-movie-app-data-mock/app/build.gradle [hunk 21]
```diff
@@ -54,15 +54,17 @@ dependencies {
     implementation 'com.google.android.material:material:1.6.1'
     implementation "androidx.compose.ui:ui:1.3.0-beta03"
     implementation "androidx.compose.material:material:1.3.0-beta03"
+    implementation "androidx.compose.material3:material3-window-size-class:1.0.0-rc01"
+    implementation "androidx.compose.material3:material3:1.0.0-rc01"
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/build.gradle:58]: Please remove this

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCategoryLabel.kt [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+    val isSelected: Boolean,
+    val categoryText: MovieCategoryLabelTextViewState
+)
+
+sealed class MovieCategoryLabelTextViewState{
+    class InputText(val text: String): MovieCategoryLabelTextViewState()
+    class ResourceText(@StringRes val textRes: Int): MovieCategoryLabelTextViewState()
+}
+
+fun onMovieCategoryLabelClick() { }
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCategoryLabel.kt]: Not needed 
In the preview just use an empty lambda

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt (hunk 17 above)
Comment (ghost) [reply android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt]: This is still not addressed
Add a parameter similar to onClick which will be used as a callback when the user clicks on the heart icon.

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+                modifier = Modifier
+                    .padding(9.dp)
+                    .size(30.dp),
+                isFavourite = true,
+                onClick = {}
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt]: Pass in the favorite click callback that will be passed into the function.

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import coil.compose.AsyncImage
+
+data class MovieCardViewState(
+    val imageUrl: String?,
+    val title: String
+)
+
+fun onMovieCardClick() { }
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/MovieCard.kt]: You can remove this
In the preview either pass in an empty lambda or don't do anything since you have a default value for onClick :)

File: android-vjestina-homework-movie-app-data-mock/app/src/main/res/values/strings.xml [hunk 25]
```diff
@@ -1,3 +1,4 @@
 <resources>
     <string name="app_name">MovieApp</string>
+    <string name="sf">Sci-Fi</string>
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/res/values/strings.xml]: Where do we need this resource?

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/CrewItem.kt [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+    val name: String,
+    val job: String,
+)
+
+@Composable
+fun CrewItem(
+    item: CrewItemViewState,
+    modifier: Modifier = Modifier
+) {
+    Column() {
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/CrewItem.kt]: pass the modifier parameter to the column here so that the caller can customise the behavior of this component

File: android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/FavoriteButton.kt [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+import androidx.compose.foundation.shape.CircleShape
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+
+fun onFavButtonClick() { }
```
Comment (ghost) [inline android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/ui/component/FavoriteButton.kt]: Not needed. 
Pass in an empty lambda in the preview

**Review by ghost — CHANGES_REQUESTED** [review]

File: android-vjestina-homework-movie-app-data-mock/app/src/main/res/values/strings.xml (hunk 25 above)
Comment (ivanseki) [reply android-vjestina-homework-movie-app-data-mock/app/src/main/res/values/strings.xml]: In MovieCategoryLabel just for testing (preview) purposes.

File: android-vjestina-homework-movie-app-data-mock/app/build.gradle (hunk 21 above)
Comment (ivanseki) [reply android-vjestina-homework-movie-app-data-mock/app/build.gradle:58]: I've asked your colleague Matej Salković about this, because I've used Compose Material 3 instead of Material Compose. He consulted with other colleagues and said it's ok to use Material 3 and that I only need to create UserScoreProgressBar with canvas and not use template from Material 3

File: android-vjestina-homework-movie-app-data-mock/app/src/main/res/values/strings.xml (hunk 25 above)
Comment (ghost) [reply android-vjestina-homework-movie-app-data-mock/app/src/main/res/values/strings.xml]: No need to extract then
Preview is used for development purposes only so you can hardcode this value in the preview function

File: android-vjestina-homework-movie-app-data-mock/app/build.gradle (hunk 21 above)
Comment (ghost) [reply android-vjestina-homework-movie-app-data-mock/app/build.gradle:58]: Awesome. Thanks for the update. Leave it as is then 🚀

**Comment by ivanseki** [comment]
Can I get a review of latest changes please?

**Review by ghost — APPROVED** [review]
Good job 💪🏻

**Comment by ghost** [comment]
Since this is approved, you can merge this pull request now
